master-firmware/src/robot_helpers/trajectory_helpers.c

```diff
@@ -1,54 +1,125 @@
 #include <ch.h>
 
+#include <timestamp/timestamp.h>
+#include <error/error.h>
+
 #include "trajectory_manager/trajectory_manager_utils.h"
+#include "base/map.h"
+#include "robot_parameters.h"
+#include "math_helpers.h"
+#include "beacon_helpers.h"
 
 #include "trajectory_helpers.h"
 
-void trajectory_wait_for_finish(struct trajectory* robot_traj)
+
+int trajectory_wait_for_end(struct _robot *robot, messagebus_t *bus, int watched_end_reasons)
 {
+#ifndef TESTS
     chThdSleepMilliseconds(100);
-    while(!trajectory_finished(robot_traj)) {
+#endif
+
+    int traj_end_reason = 0;
+    while(traj_end_reason == 0) {
+        traj_end_reason = trajectory_has_ended(robot, bus, watched_end_reasons);
+#ifndef TESTS
         chThdSleepMilliseconds(1);
+#endif
     }
+    NOTICE("End of trajectory reason %d at %d %d",
+            traj_end_reason, position_get_x_s16(&robot->pos), position_get_y_s16(&robot->pos));
+
+    return traj_end_reason;
 }
 
-void trajectory_wait_for_collision(struct blocking_detection* distance_blocking)
+int trajectory_has_ended(struct _robot *robot, messagebus_t *bus, int watched_end_reasons)
 {
-    chThdSleepMilliseconds(100);
-    while(!bd_get(distance_blocking)) {
-        chThdSleepMilliseconds(1);
+    if ((watched_end_reasons & TRAJ_END_GOAL_REACHED) && trajectory_finished(&robot->traj)) {
+        return TRAJ_END_GOAL_REACHED;
     }
+
+    if ((watched_end_reasons & TRAJ_END_COLLISION) && bd_get(&robot->distance_bd)) {
+        return TRAJ_END_COLLISION;
+    }
+
+    if ((watched_end_reasons & TRAJ_END_COLLISION) && bd_get(&robot->angle_bd)) {
+        return TRAJ_END_COLLISION;
+    }
+
+    if (watched_end_reasons & TRAJ_END_OPPONENT_NEAR) {
+        beacon_signal_t beacon_signal;
```

Comment:
True, this is a design choice for now, we'll worry about support for two opponents later. I'll create a related issue for this.